crates/apollo-mcp-server/src/apps/execution.rs

```diff
@@ -0,0 +1,291 @@
+use std::sync::Arc;
+
+use futures::TryFutureExt;
+use futures::future::try_join_all;
+use http::HeaderMap;
+use opentelemetry::Context;
+use opentelemetry::trace::FutureExt;
+use rmcp::model::{CallToolResult, Content, JsonObject};
+use serde_json::{Map, Value};
+use url::Url;
+
+use crate::errors::McpError;
+use crate::graphql::{self, Executable};
+use crate::operations::Operation;
+
+use super::{App, AppTool};
+
+pub(crate) async fn find_and_execute_app(
+    apps: &[App],
+    tool_name: &str,
+    headers: &HeaderMap,
+    arguments: Option<&JsonObject>,
+    endpoint: &Url,
+) -> Option<Result<CallToolResult, McpError>> {
+    for app in apps {
+        for tool in &app.tools {
+            if tool.tool.name == tool_name {
+                return Some(execute_app(app, tool, headers, arguments, endpoint).await);
+            }
+        }
+    }
+    None
+}
+
+pub(crate) async fn execute_app(
+    app: &App,
+    tool: &AppTool,
+    headers: &HeaderMap,
+    arguments: Option<&JsonObject>,
+    endpoint: &Url,
+) -> Result<CallToolResult, McpError> {
+    let graphql_request = graphql::Request {
+        input: Value::from(filter_inputs_for_operation(arguments, &tool.operation)),
+        endpoint,
+        headers,
+    };
+
+    let result = tool
+        .operation
+        .execute(graphql_request)
+        .with_context(Context::current())
+        .await?;
+
+    let mut prefetch_calls = Vec::new();
+    for prefetch in &app.prefetch_operations {
+        if Arc::ptr_eq(&prefetch.operation, &tool.operation) {
+            // Don't re-run any prefetches that already ran as the base tool call
+            continue;
+        }
+
+        let graphql_request = graphql::Request {
+            input: Value::from(filter_inputs_for_operation(arguments, &prefetch.operation)),
+            endpoint,
+            headers,
+        };
+        prefetch_calls.push(
+            prefetch
+                .operation
+                .execute(graphql_request)
+                .with_context(Context::current())
+                .map_ok(|res| (prefetch.prefetch_id.clone(), res)),
+        );
+    }
+
+    let prefetch_results = try_join_all(prefetch_calls.into_iter()).await?;
```

Comment:
With `try_join_all`, it looks like if any prefetch operation fails, the entire tool call fails. Is this the desired behavior?